# Анализ базы данных книг

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель проекта:** проанализировать базу данных и подготовить предложение для нового продукта

## Описание данных

**Таблица books**

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

![Схема БД](https://pictures.s3.yandex.net/resources/scheme_1589269096.png "Схема БД")

## Подключение к БД

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Ознакомление с данными

In [3]:
# напишем функцию для более простого ввода запросов к БД
def query(query_db):
    return pd.io.sql.read_sql(query_db, con = engine)

**Таблица books**

In [4]:
query_select_1 = ''' 
SELECT *
FROM books;
'''
query(query_select_1)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

**Таблица authors**

In [5]:
query_select_2 = ''' 
SELECT *
FROM authors;
'''
query(query_select_2)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

**Таблица publishers**

In [6]:
query_select_3 = ''' 
SELECT *
FROM publishers;
'''
query(query_select_3)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

**Таблица ratings**

In [7]:
query_select_4 = ''' 
SELECT *
FROM ratings;
'''
query(query_select_4)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

**Таблица reviews**

In [8]:
query_select_5 = ''' 
SELECT *
FROM reviews;
'''
query(query_select_5)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В базе данных содержится информация о 1000 книгах, 636 авторах, 340 издательствах, 6456 оценках книг и о 2793 обзорах.

## Анализ данных

**Посчитаем, сколько книг вышло после 1 января 2000 года**

In [10]:
query_count_books = ''' 
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01';
'''

query(query_count_books)

count
0    819

***Вывод:*** 

После 1 января 2000 года вышло 819 книг

**Для каждой книги посчитаем количество обзоров и среднюю оценку**

In [11]:
query_count_books_1 = ''' 
SELECT b.book_id, b.title, COUNT(DISTINCT rev.review_id) AS count_review, AVG(rat.rating) AS mean_rating
FROM books AS b 
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS rat ON rev.book_id = rat.book_id
GROUP BY b.book_id
ORDER BY count_review DESC
LIMIT 10;
'''

query(query_count_books_1)

book_id                                              title  count_review  \
0      948                            Twilight (Twilight  #1)             7   
1      696             The Da Vinci Code (Robert Langdon  #2)             6   
2      207                                    Eat  Pray  Love             6   
3      627                                      The Alchemist             6   
4      673                             The Catcher in the Rye             6   
5      695  The Curious Incident of the Dog in the Night-Time             6   
6      302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
7      299  Harry Potter and the Chamber of Secrets (Harry...             6   
8      497                          Outlander (Outlander  #1)             6   
9      656                                     The Book Thief             6   

   mean_rating  
0     3.662500  
1     3.830508  
2     3.395833  
3     3.789474  
4     3.825581  
5     4.081081  
6     4.414634  
7     4.287500  
8     4.125000  
9     4.264151

In [12]:
query_count_books_2 = ''' 
SELECT b.book_id, b.title, COUNT(DISTINCT rev.review_id) AS count_review, AVG(rat.rating) AS mean_rating
FROM books AS b 
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS rat ON rev.book_id = rat.book_id
GROUP BY b.book_id
ORDER BY mean_rating DESC;
'''

query(query_count_books_2)

book_id                                              title  count_review  \
0        672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
1        808                            The Natural Way to Draw             0   
2         83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
3        387                               Leonardo's Notebooks             0   
4        191  Disney's Beauty and the Beast (A Little Golden...             0   
..       ...                                                ...           ...   
995      915  The World Is Flat: A Brief History of the Twen...             3   
996      371                                              Junky             2   
997      202                                      Drowning Ruth             3   
998      316                  His Excellency: George Washington             2   
999      303                               Harvesting the Heart             2   

     mean_rating  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
..           ...  
995         2.25  
996         2.00  
997         2.00  
998         2.00  
999         1.50  

[1000 rows x 4 columns]

***Вывод:***

В нашей базе максимальное количество обзоров (7) пришлось на книгу "Twilight" ("Сумерки"). 
Есть книги, по которым не было написано ни одного обзора и нет оценок.

**Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры**

In [13]:
query_publisher = ''' 
SELECT p.publisher, p.count_books
FROM (SELECT b.publisher_id AS publisher_id,
      COUNT(b.book_id) AS count_books,
      publishers.publisher AS publisher 
      FROM books AS b
      LEFT JOIN publishers ON b.publisher_id=publishers.publisher_id
      WHERE num_pages>50
      GROUP BY b.publisher_id, publishers.publisher
      ORDER BY count_books DESC) AS p
LIMIT 1;
'''

query(query_publisher)

publisher  count_books
0  Penguin Books           42

***Вывод:*** 

Мы нашли издательство, которое выпустило наибольшее число книг толще 50 страниц. Это издательство называется "Penguin Books"

**Определим автора с самой высокой средней оценкой книг (учитывать будем только книги с 50 и более оценками)**

In [14]:
query_author = ''' 
SELECT a.author, AVG(r.rating) AS mean_rating
FROM books AS b 
JOIN ratings AS r ON b.book_id = r.book_id
JOIN authors AS a ON b.author_id = a.author_id
WHERE b.book_id IN (
                     SELECT book_id
                     FROM ratings
                     GROUP BY book_id
                     HAVING COUNT(rating_id) >= 50
                    )
GROUP BY a.author
ORDER BY mean_rating DESC
LIMIT 1;

'''

query(query_author)

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.287097

***Вывод:*** 

Автором с самой высокой средней оценкой книг является Джоан Роулинг (J.K. Rowling/Mary GrandPré), написавшая романы о Гарри Поттере

**Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок**

In [15]:
query_mean_reviews = ''' 
SELECT AVG(count_review) AS mean_count_review
FROM (
        SELECT rev.username, COUNT(DISTINCT rev.review_id) AS count_review
        FROM reviews AS rev
        WHERE rev.username IN (SELECT username
                               FROM ratings
                               GROUP BY username
                               HAVING COUNT(DISTINCT rating_id) > 50)
                               GROUP BY username) AS u 

'''

query(query_mean_reviews)

mean_count_review
0          24.333333

***Вывод:*** 

Активные пользователи, которые поставили более 50 оценок, в среднем написали по 24 обзора.

## Выводы

- После 1 января 2000 года было выпущено 819 книг
- В нашей базе максимальное количество обзоров (7) пришлось на книгу "Twilight" ("Сумерки"). Есть книги, по которым не было написано ни одного обзора и нет оценок
- Издательство, выпустившее наибольшее число книг - "Penguin Books"
- Автором с самой высокой средней оценкой книг является Джоан Роулинг (J.K. Rowling/Mary GrandPré), написавшая романы о Гарри Поттере
- Активные пользователи, которые поставили более 50 оценок, в среднем написали по 24 обзора.
